{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "\n",
    "openai_api_key = ''\n",
    "client = OpenAI(api_key=openai_api_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_manifest(wrapper_script_fp, LSID, author, docker_image, repo, documentation_url, filepath, output_fp='output/manifest'):\n",
    "    with open(wrapper_script_fp, 'r') as file:\n",
    "        wrapper_script = file.read()\n",
    "    \n",
    "    completion = client.chat.completions.create(\n",
    "    model=\"ft:gpt-4o-mini-2024-07-18:personal::B1TFUey0\",\n",
    "    messages=[\n",
    "        {\"role\": \"system\", \"content\": \"\"\"You are a genius GenePattern developer who writes manifest files based on wrapper scripts. You\n",
    "will be fired for any mistakes you make\"\"\"},\n",
    "        {\"role\": \"user\", \"content\": f\"\"\"\n",
    "Act as a senior software developer. I am going to describe a GenePattern manifest file, and then I would like you to create one based on information I provide. The first line is a \"#\" followed by the name of the function I provide. The second line is a # followed by the current date and time. Do not include any blank lines in the output.\n",
    "The next section should be pasted verbatim:\n",
    "---------------------------------------------------------\n",
    "JVMLevel=\n",
    "LSID=\n",
    "author=GenePattern Team + ChatGPT\n",
    "commandLine=\n",
    "cpuType=any\n",
    "The next line should be \"description=\" followed by a brief description of the function I provided.\n",
    "The next line should be \"documentationUrl=\" followed by the URL to the site where the function is described.\n",
    "The next line should be \"fileFormat=\" followed by a comma-separated list of the file extensions output by the function.\n",
    "The next line should be \"job.cpuCount=\"\n",
    "The next line should be \"job.docker.image=[DOCKER IMAGE HERE]\"\n",
    "The next section should be pasted verbatim:\n",
    "job.memory=\n",
    "job.walltime=\n",
    "language=any\n",
    "The next line should start with \"categories=\" and should be most applicable term in following list: alternative splicing,batch correction,clustering,cnv analysis,data format conversion,differential expression,dimension reduction,ecdna,flow cytometry,gene list selection,gsea,image creators,metabolomics,methylation,missing value imputation,mutational significance analysis,pathway analysis,pipeline,prediction,preprocess & utilities,projection,proteomics,rna velocity,rna-seq,rnai,sage,sequence analysis,single-cell,snp analysis,statistical methods,survival analysis,variant annotation,viewer,visualizer\n",
    "The next line should be \"name=\" and then the name of the function.\n",
    "The next line should be \"os=any\"\n",
    "After that, identify all the parameters in the wrapper script. \n",
    "---------------------------------------------------------\n",
    "The manifest file should include each parameter of the provided function, in the format. I will provide below. Here are some instructions for this format:\n",
    "1. When you see a # character, replace it with the number of the parameter in the provided function.\n",
    "2. When you see \"default_value=\", place the parameter's default value after the \"=\" if there is one.\n",
    "3. When you see \"description=\", add the parameter's description after the \"=\"\n",
    "4. When you see \"name=\", add the parameter's name after the \"=\"\n",
    "5. When you see \"optional=\", write \"on\" if the parameter is optional\n",
    "6. In the parameter name, replace the \"#\" with the cardinal number of the parameter\n",
    "7. When you see \"flag=\", add the parameter's command-line flag after the \"=\" if there is one. If there are more than one way of specifying a flag, use the one that starts with two hyphens: \"--\"\n",
    "8. When you see \"type=\", add the parameter's type after the \"=\". The type should be the term in the following comma-separated list that corresponds most closely to the type of parameter: CHOICE,FILE,Floating Point,Integer,TEXT,java.lang.String. Pick java.io.File if you think the input are filepaths. \n",
    "9. When you see \"taskType=\", add the type of analysis this module performs. For example: batch correction, visualizer, scRNA analysis. You can infer the category based on the name and description.\n",
    "10 If you think a parameter is a file path, put IN for the p#_MODE\n",
    "11. Order from p1-pn\n",
    "12. Enter a new line between parameter info (ex: new line between p1_... and p2_...)\n",
    "\n",
    "---------------------------------------------------------\n",
    "Here is the format for each parameter:\n",
    "p#_MODE=\n",
    "p#_TYPE=\n",
    "p#_default_value=\n",
    "p#_description=\n",
    "p#_fileFormat=\n",
    "p#_flag=\n",
    "p#_name=\n",
    "p#_numValues=\n",
    "p#_optional=\n",
    "p#_prefix=\n",
    "p#_prefix_when_specified=\n",
    "p#_type=\n",
    "p#_value=\n",
    "taskType= \n",
    "---------------------------------------------------------\n",
    "For the commandline, generate a Rscript commandline to run the wrapper script. The parameters should be: --flag <value>. \n",
    "\n",
    "Relevant info:\n",
    "LSID: {LSID}\n",
    "author: {author}\n",
    "docker_image: {docker_image}\n",
    "repo: {repo}\n",
    "documentation_url: {documentation_url},\n",
    "filepath: {filepath}\n",
    "wrapper_script: {wrapper_script}\"\"\"}\n",
    "    ])\n",
    "\n",
    "    manifest = completion.choices[0].message.content\n",
    "    \n",
    "    with open(output_fp, 'w') as file:\n",
    "        file.write(manifest)\n",
    "    file.close()\n",
    "    \n",
    "    return manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = create_manifest(wrapper_script_fp='input/wrapper.R',\n",
    "                LSID='urn:lsid:genepattern.org:module.analysis:00465:999999999',\n",
    "                author='Thorin Tabor;UCSD - Mesirov Lab',\n",
    "                docker_image='genepattern/spatialge-stgradient:0.2',\n",
    "                repo='https://github.com/genepattern/spatialGE.STgradient',\n",
    "                documentation_url='https://genepattern.github.io/spatialGE.STgradient/v1/',\n",
    "                filepath='/spatialGE/wrapper.R',)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "3.12.7",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
